ampd/src/solana/verifier_set_verifier.rs

```diff
@@ -0,0 +1,285 @@
+use solana_transaction_status::EncodedConfirmedTransactionWithStatusMeta;
+use thiserror::Error;
+use tracing::error;
+
+use gmp_gateway::events::GatewayEvent;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum VerificationError {
+    #[error("Failed to parse tx log messages")]
+    NoLogMessages,
+    #[error("Tried to get gateway event from program logs, but couldn't find anything.")]
+    NoGatewayEventFound,
+}
+
+type Result<T> = std::result::Result<T, VerificationError>;
+
+pub fn parse_gateway_event(tx: &EncodedConfirmedTransactionWithStatusMeta) -> Result<GatewayEvent> {
+    let Some(meta) = &tx.transaction.meta else {
+        return Err(VerificationError::NoLogMessages);
+    };
+
+    let log_messages = match &meta.log_messages {
+        solana_transaction_status::option_serializer::OptionSerializer::Some(log_msg) => log_msg,
+        _ => return Err(VerificationError::NoLogMessages),
+    };
+
+    log_messages
+        .iter()
+        .find_map(GatewayEvent::parse_log)
+        .ok_or(VerificationError::NoGatewayEventFound)
+}
+
+#[tracing::instrument(name = "sol_verify_verifier_set")]
+pub fn verify_verifier_set(
+    verifier_set_conf: &VerifierSetConfirmation,
+    signers: &Vec<Address>,
+    weights: &Vec<u128>,
+    quorum: u128,
+) -> Vote {
+    let tx_id = &verifier_set_conf.tx_id;
+    let verifier_set = &verifier_set_conf.verifier_set;
+    let verifier_set_threshold = verifier_set.threshold.u128();
+
+    if signers.len() != weights.len() {
+        error!(
+            tx_id,
+            "signers length do not match in sol data: signers len() was {}, weights len() was {}",
+            signers.len(),
+            weights.len()
+        );
+        return Vote::FailedOnChain;
+    }
+
+    if verifier_set_threshold != quorum {
+        error!(
+            tx_id,
+            "threshold do not match: axelar was {}, solana was {}", verifier_set.threshold, quorum
+        );
+        return Vote::FailedOnChain;
+    }
+
+    for (sol_addr, sol_weight) in signers.iter().zip(weights.iter()) {
+        let sol_addr = sol_addr.encode_hex::<String>();
+        let Some((addr, signer)) = verifier_set.signers.get_key_value(&sol_addr) else {
+            error!(
+                tx_id,
+                "Lookup for sol signer failed on axelar data, missing sol address was: {}",
+                sol_addr
+            );
+            return Vote::FailedOnChain;
+        };
+        let signer_address = signer.address.to_string();
+        if *addr != signer_address {
+            error!(
+                tx_id,
+                "Axelar data seems corrupted. Map key seems different than signer address. Map key was {}, while signer address was {}.",
+                addr,
+                signer_address
+            );
+            return Vote::FailedOnChain;
+        }
+        let signer_pub_key = signer.pub_key.encode_hex::<String>();
+        if sol_addr != signer_pub_key {
```

Comment:
Is this ok ? This variables are the hexadecimal representation of the [solana address](https://github.com/eigerco/axelar-amplifier/blob/ef4f79e18e74524e0824d52f56415007dd4e7639/contracts/multisig/src/msg.rs#L78-L79) and the signer [PublicKey type](https://github.com/eigerco/axelar-amplifier/blob/ef4f79e18e74524e0824d52f56415007dd4e7639/contracts/multisig/src/key.rs#L118-L119) respectively.

Both types seems to hold the raw bytes, which then are converted into its hexadecimal representation ([1](https://github.com/eigerco/axelar-amplifier/pull/24/files#diff-8ce2329368b450c7d889a373599fab6cc28aa8a62dd4b028f0ddc21f314dbf93R68), [2](https://github.com/eigerco/axelar-amplifier/pull/24/files#diff-8ce2329368b450c7d889a373599fab6cc28aa8a62dd4b028f0ddc21f314dbf93R87)) for comparison here.

Reply:
I have forgotten the context. What does this function do? What are we trying to compare here? When does this get called? 🤔 

But overall the comparisons seem to make sense. And it should also work after we migrate to the rkyv encoding scheme (meaning that the Solana Gateway will also support ed25519 pubkeys)

Reply:
This function is the core logic that compares the new proposed VerifierSet with the one verified on the solana blockchain. Basically we do the same checking as with the message (wait for an axelar event, compare against integrated chain), but for a different data structure. We check the incoming new `VerifierSet` from axelar is the same as the published one on the Solana network and we emit the proper vote.

Reply:
The types do not facilitates this, but i fixed it at 2f463cb1034d517c48a1c5f42ff14cedc6d23316 . I wonder if the Axelar's [PublicKey](https://github.com/eigerco/axelar-amplifier/blob/2f463cb1034d517c48a1c5f42ff14cedc6d23316/contracts/multisig/src/key.rs#L116-L125) enum should implement a way of getting the inner type `HexBinary` of each respective variant transparently, for this cases in which we do not mind the kind of key. Similar to this:

```rust
#[cw_serde]
#[derive(Ord, PartialOrd, Eq)]
pub enum PublicKey {
    /// ECDSA public key must be in compressed format (33 bytes)
    #[serde(deserialize_with = "deserialize_ecdsa_key")]
    Ecdsa(HexBinary),

    #[serde(deserialize_with = "deserialize_ed25519_key")]
    Ed25519(HexBinary),
}

// Here implement a way to facilitate the inner HexBinary without having to "match" in all the rest of the code.
impl PublicKey {
    pub fn hex_binary(&self) -> &HexBinary {
        match self {
            PublicKey::Ecdsa(hb) | PublicKey::Ed25519(hb) => hb,
        }
    }
}
```

I just wonder if this could have other non desired side effects. I am one step from doing and upstream PR (directly from `main`) with this change guys.

Reply:
Seems this could be a larger discussion. I am going to create an issue for continuing the discussion there.